Fast DDS CMake options documentation (#156)
eProsima/Fast-RTPS-docs: docs/installation/configuration/cmake_options.rst

files, for later translation into C++ source code.
These libraries can also be configured using Fast DDS CMake options.

.. list-table::
    :header-rows: 1

    *   - Option
        - Description
        - Default
    *   - :class:`THIRDPARTY`
        - Activates the use of Fast CDR and the internal third-party libraries. It is set to ``ON`` |br|
          if EPROSIMA_BUILD is ``ON``.
        - ``OFF``
    *   - :class:`THIRDPARTY_UPDATE`
        - Activates the automatic update of Fast CDR and the internal third-party libraries.
        - ``ON``
    *   - :class:`THIRDPARTY_fastcdr`
        - Activates the use of Fast CDR library.
        - ``OFF``
    *   - :class:`THIRDPARTY_Asio`
        - Activates the use of Asio internal third-party library.
        - ``OFF``
    *   - :class:`THIRDPARTY_TinyXML2`
        - Activates the use of TinyXML2 internal third-party library.
        - ``OFF``
    *   - :class:`THIRDPARTY_android-ifaddrs`
        - Activates the use of android-ifaddrs, an implementation of :func:`getifaddrs` for Android. |br|
          Only if :class:`ANDROID` is set to ``ON``.
        - ``OFF``

> Test options

Reviewer: As a general comment for the subsection, I think we could combine all the options into a single table. What do you think?

------------

> *eProsima Fast DDS* comes with a full set of tests to check the performance of the library.

Reviewer: It's not only for performance testing. In fact, 99+ % of the tests are for continuous integration. There are basically the following types of tests:
* Unit tests
* Blackbox tests
* Performance tests
* Profiling
* XTypes

> These tests are divided according to the layer being tested; thus there are DDS-layer tests and RTPS-layer tests.

Reviewer: This does not apply to all test types from above

The building and execution of these tests is specified by the Fast DDS CMake options shown in the table below.

.. list-table::
    :header-rows: 1

    *   - Option
        - Description
        - Default
    *   - :class:`FASTRTPS_API_TESTS`
        - Enables the execution of black-box tests for Fast DDS using the Fast DDS RTPS-layer API.
        - ``OFF``
    *   - :class:`FASTDDS_PIM_API_TESTS`
        - Enables the execution of black-box tests for Fast DDS using the Fast DDS DDS-layer API.
        - ``OFF``

Fast DDS tests are implemented using the GoogleTest framework.
To run the tests it is required to enable this framework using the :class:`GTEST_INDIVIDUAL` option.
The configuration of the installation directories is also allowed as indicated in the following table.

.. list-table::
    :header-rows: 1

    *   - Option
        - Description
        - Default
    *   - :class:`GTEST_INDIVIDUAL`

>         - Activate the execution of GoogleTest tests.

Reviewer: No exactly. It activates the individual building of the tests. However, the test are compiled if `EPROSIMA_BUILD=ON` (not 100% sure that's the correct option). In that case, if `GTEST_INDIVIDUAL=OFF`, then when you run the tests, they are one huge test, instead of 1000+ individual ones.

        - ``OFF``
    *   - :class:`EPROSIMA_GTEST`
        - Activates special set of GTEST_ROOT (the root directory of the GoogleTest installation).
        - ``OFF``
    *   - :class:`EPROSIMA_GMOCK`

>         - Activates special set of GMOCK_ROOT (the root directory of the GoogleMock installation).

Reviewer: I'm not sure what this means

>         - ``OFF``

Reviewer: I'd put this table as the first one

Furthermore, several CMake options are available to specify which tests are to be executed.

.. list-table::
    :header-rows: 1

    *   - Option
        - Description
        - Default
    *   - :class:`PERFORMANCE_TESTS`

>         - Activates the building and execution of performance tests.

Reviewer: Except for the video one. For that one, you need to set both `PERFORMANCE_TESTS=ON` and `VIDEO_TESTS=ON`

>         - Activates the building and execution of performance tests.

Reviewer: In general, this activates the building. Of course that means that you can run them, but I think we don't need to say that.
```suggestion
        - Activates the building of performance tests.
```

        - ``OFF``
    *   - :class:`PROFILING_TESTS`
        - Activates the building and execution of profiling tests using Valgrind.
        - ``OFF``
    *   - :class:`EPROSIMA_BUILD_TESTS`
        - Activates the building and execution of black-box, unit, xtypes, RTPS communication and |br|
          DDS communication tests. It is set to ``ON`` if :class:`EPROSIMA_BUILD` is ``ON`` and
          :class:`EPROSIMA_INSTALLER` |br| is ``OFF``.
        - ``OFF``
    *   - :class:`VIDEO_TESTS`

>         - Activate the building and execution of video performance tests.

Reviewer: Only if `PERFORMANCE_TESTS` is also `ON`

General options
^^^^^^^^^^^^^^^

The *Fast DDS* CMake options for configuring general settings are shown below, together with their description and
dependency on other options.

.. list-table::
    :header-rows: 1

    *   - Option
        - Description
        - Default
    *   - :class:`EPROSIMA_INSTALLER`
        - Creates a build for Windows binary installers. Specifically it adds to the list of |br|
          components to install (:class:`CPACK_COMPONENTS_ALL`) the libraries corresponding |br|
          to the Microsoft Visual C++ compiler (MSVC). Setting :class:`EPROSIMA_INSTALLER` |br|
          to ``ON`` has the following effects on other options:

            - :class:`EPROSIMA_BUILD` is set to ``ON``.
            - :class:`BUILD_DOCUMENTATION` is set to ``ON``.
            - :class:`INSTALL_EXAMPLES` is set to ``ON``.

        - ``OFF``
    *   - :class:`EPROSIMA_BUILD`
        - Activates internal *Fast DDS* builds.
          It is set to ``ON`` if :class:`EPROSIMA_INSTALLER` is ``ON``. |br|
          Setting :class:`EPROSIMA_BUILD` to ``ON`` has the following effects on other options:

            - :class:`INTERNAL_DEBUG` is set to ``ON``.

>             - :class:`SHM_TRANSPORT_DEFAULT` is set to ``ON`` and :class:`EPROSIMA_INSTALLER` is set to OFF.

Reviewer: ```suggestion
            - :class:`SHM_TRANSPORT_DEFAULT` is set to ``ON`` and :class:`EPROSIMA_INSTALLER` is set to ``OFF``.
```

General options
^^^^^^^^^^^^^^^

The *Fast DDS* CMake options for configuring general settings are shown below, together with their description and
dependency on other options.

.. list-table::
    :header-rows: 1

    *   - Option
        - Description
        - Default
    *   - :class:`EPROSIMA_INSTALLER`
        - Creates a build for Windows binary installers. Specifically it adds to the list of |br|
          components to install (:class:`CPACK_COMPONENTS_ALL`) the libraries corresponding |br|
          to the Microsoft Visual C++ compiler (MSVC). Setting :class:`EPROSIMA_INSTALLER` |br|
          to ``ON`` has the following effects on other options:

            - :class:`EPROSIMA_BUILD` is set to ``ON``.
            - :class:`BUILD_DOCUMENTATION` is set to ``ON``.
            - :class:`INSTALL_EXAMPLES` is set to ``ON``.

        - ``OFF``
    *   - :class:`EPROSIMA_BUILD`
        - Activates internal *Fast DDS* builds.
          It is set to ``ON`` if :class:`EPROSIMA_INSTALLER` is ``ON``. |br|
          Setting :class:`EPROSIMA_BUILD` to ``ON`` has the following effects on other options:

            - :class:`INTERNAL_DEBUG` is set to ``ON``.

>             - :class:`SHM_TRANSPORT_DEFAULT` is set to ``ON``.

Reviewer: Update for the [discovered bug](https://github.com/eProsima/Fast-DDS/pull/1249)
```suggestion
            - :class:`SHM_TRANSPORT_DEFAULT` is set to ``ON`` and `EPROSIMA_INSTALLER` is set to ``OFF``.
```

          to create shared libraries if on. All libraries are built shared unless the library was |br|
          explicitly added as a static library.
        - ``ON``
    *   - :class:`SECURITY`
        - Activates the *Fast DDS* security module. Please refer to :ref:`security` for more information |br|
          on security module.
        - ``OFF``
    *   - :class:`NO_TLS`
        - Disables Transport Layer Security (TLS) Support. Please refer to :ref:`transport_tcp_tls` for |br|
          more information on *Fast DDS* TLS configuration.
        - ``OFF``
    *   - :class:`SHM_TRANSPORT_DEFAULT`
        - Adds Shared Memory transport (SHM) to the default transports. It is set to ``ON`` if |br|
          :class:`EPROSIMA_BUILD` is ``ON``. Please refer to :ref:`SHM <transport_sharedMemory_sharedMemory>` section
          for more information |br|
          on *Fast DDS* SHM transport.
        - ``OFF``
    *   - :class:`COMPILE_EXAMPLES`
        - Builds the *Fast DDS* examples. It is set to ``ON`` if :class:`EPROSIMA_BUILD` is ``ON`` and |br|
          :class:`EPROSIMA_INSTALLER` is ``OFF``. These examples can be found in the

>           `eProsima *Fast DDS* <https://github.com/eProsima/Fast-DDS/tree/master/examples>`_ |br|

Reviewer: Italics don't work inside a reference
```suggestion
          `eProsima Fast DDS <https://github.com/eProsima/Fast-DDS/tree/master/examples>`_ |br|
```

        - ``OFF``
    *   - :class:`INTERNAL_DEBUG`
        - Activates |Log::Kind::Info| debug messages (See :ref:`dds_layer_log_intro`). |br|
          For this option to have any effect, i.e. to print the information messages,  |br|
          *Fast DDS* must be debug built. This is done by setting the :class:`CMAKE_BUILD_TYPE` |br|
          option to `Debug`. Moreover, :class:`INTERNAL_DEBUG` is set to ``ON`` if |br|
          :class:`EPROSIMA_BUILD` is ``ON``.
        - ``OFF``

Third-party libraries options
-----------------------------

*Fast DDS* relies on the `eProsima FastCDR <https://github.com/eProsima/Fast-CDR>`_ library for serialization
mechanisms.
Moreover, *Fast DDS* requires two external dependencies for its proper operation: Asio and TinyXML2.
Asio is a cross-platform C++ library for network and low-level I/O programming, while TinyXML2 parses the XML profile
files, so *Fast DDS* can use them (see :ref:`xml_profiles`).
These three libraries (eProsima FastCDR, Asio and TinyXML2) can be installed by the user, or downloaded on the
*Fast DDS* build.
In the latter case, they are referred to as *Fast DDS* internal third-party libraries.

> This can be done by setting :class:`THIRDPARTY` to ``ON`` or :class:`EPROSIMA_BUILD` is set to ``ON``.

Reviewer: ```suggestion
This can be done by setting either :class:`THIRDPARTY` or :class:`EPROSIMA_BUILD` to ``ON``.
```

.. list-table::
    :header-rows: 1

    *   - Option
        - Description
        - Default
    *   - :class:`THIRDPARTY`
        - Activates the use of Fast CDR and the internal third-party libraries. It is set to ``ON`` |br|
          if :class:`EPROSIMA_BUILD` is ``ON``.
        - ``OFF``
    *   - :class:`THIRDPARTY_UPDATE`
        - Activates the automatic update of just Fast CDR library.
        - ``ON``
    *   - :class:`THIRDPARTY_fastcdr`
        - Links against the internal Fast CDR library.
        - ``OFF``
    *   - :class:`THIRDPARTY_Asio`
        - Links against the internal Asio internal third-party library.
        - ``OFF``
    *   - :class:`THIRDPARTY_TinyXML2`
        - Links against the internal TinyXML2 internal third-party library.
        - ``OFF``
    *   - :class:`THIRDPARTY_android-ifaddrs`
        - Links against the internal android-ifaddrs, an implementation of :func:`getifaddrs` |br|
          for Android. Only if :class:`ANDROID` is ``ON``.
        - ``OFF``

.. Note::

    :class:`ANDROID` is a CMake environment variable that is set to ``1`` if the target system
    (:class:`CMAKE_SYSTEM_NAME`) is Android.

Test options
------------

> *eProsima *Fast DDS** comes with a full set of tests for continuous integration.

Reviewer: ```suggestion
*eProsima Fast DDS* comes with a full set of tests for continuous integration.
```

The types of tests are: unit tests, black-box tests, performance tests, profiling tests, and
XTypes tests.
The building and execution of these tests is specified by the *Fast DDS* CMake options shown in the table below.

Moreover, *Fast DDS* tests are implemented using the GoogleTest framework.
To run the tests it is required to enable this framework using the :class:`GTEST_INDIVIDUAL` option.

> The configuration of the installation directories is also allowed as indicated in the following table.

Reviewer: The table is more correct than this
